From 3fd18662daf67e4b86e9b18d838c08fc45d427dc Mon Sep 17 00:00:00 2001
From: David Faure <david.faure@kdab.com>
Date: Tue, 8 Sep 2026 15:03:20 +0200
Subject: [PATCH] QMimeDatabase: fix parsing of hex escapes in magic values
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

makePattern consumed the character after a one-digit hex escape such as
\xa even when it wasn't a hex digit. shared-mime-info 2.5 uses
"\xa\x0\x0\x0MSWordDoc\x0" for application/msword, which turned into a
newline followed by the literal "x0", so the rule never matched.
Only consume hex digits, and treat \x without any digit as a literal
'x', like update-mime-database does, so that the XML and cache
providers agree.

Pick-to: 6.11
Change-Id: I05d558c104810e0fa956d036cee3ad4de42ba849
Reviewed-by: Mårten Nordheim <marten.nordheim@qt.io>
Reviewed-by: Tatiana Borisova <tatiana.borisova@qt.io>
(cherry picked from commit 3b4383877b8e65051bb6ea9e1a1304f531257c0d)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/3fd18662daf67e4b86e9b18d838c08fc45d427dc
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/corelib/mimetypes/qmimemagicrule.cpp             | 12 ++++++------
 .../mimetypes/qmimedatabase/hex-escape-magic.xml     | 11 +++++++++++
 .../qmimedatabase-cache-builtin/CMakeLists.txt       |  1 +
 .../qmimedatabase-cache-fdoxml/CMakeLists.txt        |  1 +
 .../qmimedatabase-xml-builtin/CMakeLists.txt         |  1 +
 .../qmimedatabase-xml-fdoxml/CMakeLists.txt          |  1 +
 .../mimetypes/qmimedatabase/tst_qmimedatabase.cpp    |  5 +++++
 7 files changed, 26 insertions(+), 6 deletions(-)
 create mode 100644 tests/auto/corelib/mimetypes/qmimedatabase/hex-escape-magic.xml

diff --git a/src/corelib/mimetypes/qmimemagicrule.cpp b/src/corelib/mimetypes/qmimemagicrule.cpp
index aba11f54450..91a5f58eb85 100644
--- a/src/corelib/mimetypes/qmimemagicrule.cpp
+++ b/src/corelib/mimetypes/qmimemagicrule.cpp
@@ -157,14 +157,14 @@ static inline QByteArray makePattern(const QByteArray &value)
     const char *e = p + value.size();
     for ( ; p < e; ++p) {
         if (*p == '\\' && ++p < e) {
-            if (*p == 'x') { // hex (\\xff)
-                char c = 0;
+            if (*p == 'x') { // hex (\\xff), up to 2 digits; a literal 'x' if none
+                char c = 'x';
                 for (int i = 0; i < 2 && p + 1 < e; ++i) {
+                    const int h = fromHex(p[1]);
+                    if (h == -1)
+                        break;
                     ++p;
-                    if (const int h = fromHex(*p); h != -1)
-                        c = (c << 4) + h;
-                    else
-                        continue;
+                    c = i == 0 ? char(h) : char((c << 4) + h);
                 }
                 *data++ = c;
             } else if (isOctalDigit(*p)) { // oct (\\7, or \\77, or \\377)
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/hex-escape-magic.xml b/tests/auto/corelib/mimetypes/qmimedatabase/hex-escape-magic.xml
new file mode 100644
index 00000000000..ba60b4bc573
--- /dev/null
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/hex-escape-magic.xml
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mime-info xmlns='http://www.freedesktop.org/standards/shared-mime-info'>
+    <mime-type type="application/x-hex-escape">
+        <comment>Magic value with single-digit hex escapes</comment>
+        <magic>
+            <match type="string" value="\xa\x0\x0\x0HEX" offset="0"/>
+            <!-- \x without hex digit is a literal x, like in update-mime-database -->
+            <match type="string" value="\xZ\x4A\x4a\x4\nHEX2" offset="0"/>
+        </magic>
+    </mime-type>
+</mime-info>
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-builtin/CMakeLists.txt b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-builtin/CMakeLists.txt
index 2994dd64cec..b52bd436de1 100644
--- a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-builtin/CMakeLists.txt
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-builtin/CMakeLists.txt
@@ -30,6 +30,7 @@ qt_internal_add_test(tst_qmimedatabase-cache-builtin
 set(testdata_resource_files
     "../add-extension.xml"
     "../circular-inheritance.xml"
+    "../hex-escape-magic.xml"
     "../invalid-magic1.xml"
     "../invalid-magic2.xml"
     "../invalid-magic3.xml"
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-fdoxml/CMakeLists.txt b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-fdoxml/CMakeLists.txt
index 3a3af6cc212..b4d16d122c0 100644
--- a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-fdoxml/CMakeLists.txt
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-cache-fdoxml/CMakeLists.txt
@@ -32,6 +32,7 @@ set(mimetypes_resource_files
 set(testdata_resource_files
     "../add-extension.xml"
     "../circular-inheritance.xml"
+    "../hex-escape-magic.xml"
     "../invalid-magic1.xml"
     "../invalid-magic2.xml"
     "../invalid-magic3.xml"
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-builtin/CMakeLists.txt b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-builtin/CMakeLists.txt
index 1253afb6ef8..ebf454c54f5 100644
--- a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-builtin/CMakeLists.txt
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-builtin/CMakeLists.txt
@@ -30,6 +30,7 @@ qt_internal_add_test(tst_qmimedatabase-xml-builtin
 set(testdata_resource_files
     "../add-extension.xml"
     "../circular-inheritance.xml"
+    "../hex-escape-magic.xml"
     "../invalid-magic1.xml"
     "../invalid-magic2.xml"
     "../invalid-magic3.xml"
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-fdoxml/CMakeLists.txt b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-fdoxml/CMakeLists.txt
index b09b29ff8be..6bbc57fb2f2 100644
--- a/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-fdoxml/CMakeLists.txt
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/qmimedatabase-xml-fdoxml/CMakeLists.txt
@@ -32,6 +32,7 @@ set(mimetypes_resource_files
 set(testdata_resource_files
     "../add-extension.xml"
     "../circular-inheritance.xml"
+    "../hex-escape-magic.xml"
     "../invalid-magic1.xml"
     "../invalid-magic2.xml"
     "../invalid-magic3.xml"
diff --git a/tests/auto/corelib/mimetypes/qmimedatabase/tst_qmimedatabase.cpp b/tests/auto/corelib/mimetypes/qmimedatabase/tst_qmimedatabase.cpp
index 4a5160e6e44..3353044502c 100644
--- a/tests/auto/corelib/mimetypes/qmimedatabase/tst_qmimedatabase.cpp
+++ b/tests/auto/corelib/mimetypes/qmimedatabase/tst_qmimedatabase.cpp
@@ -36,6 +36,7 @@ static const std::array additionalGlobalMimeFiles = {
     "yast2-metapackage-handler-mimetypes.xml",
     "qml-again.xml",
     "magic-and-hierarchy.xml",
+    "hex-escape-magic.xml",
 };
 
 static const std::array additionalLocalMimeFiles = {
@@ -1239,6 +1240,10 @@ void tst_QMimeDatabase::installNewGlobalMimeType()
     const QString fooTestFile2 = s_additionalFilesResourcePrefix + "magic-and-hierarchy2.foo"_L1;
     QCOMPARE(db.mimeTypeForFile(fooTestFile2).name(), QString::fromLatin1("application/vnd.qnx.bar-descriptor"));
 
+    // Hex escapes in a magic value: one digit, no digit, two digits in either case
+    QCOMPARE(db.mimeTypeForData("\n\0\0\0HEX"_ba).name(), "application/x-hex-escape"_L1);
+    QCOMPARE(db.mimeTypeForData("xZJJ\4\nHEX2"_ba).name(), "application/x-hex-escape"_L1);
+
     // Test if we can use the default comment
     {
         const auto restoreLocale = qScopeGuard([prior = QLocale()]() {
